### ISSUE TITLE
Allow CSS variables in inline styles

### ISSUE THREAD

**Issue by apottere** [OWASP/java-html-sanitizer#207]
It doesn't look like there's any way to configure the HTML sanitizer to allow css variables in inline styles.

Given the following policy:
```java
    HtmlSanitizer.Policy policy = new HtmlPolicyBuilder()
        .allowStyling()
        .allowElements("div")
        .build(renderer);
```

And the following markup:
```html
<div style="color: var(--text-inactive); padding-left: 10px;"></div>
```

The result is:
```html
<div style="padding-left:10px"></div>
```

Looking at `CssSchema`, it doesn't seem like there's even a way to create a policy that allows variables - I tried adding the following snippets:
```
    ImmutableMap<String, String> mozOutlineFunctions = ImmutableMap.of("rgb(", "rgb()", "rgba(", "rgba()", "var(", "var()");

    ...

    builder.put("var()", new Property(BIT_UNRESERVED_WORD | BIT_STRING, ImmutableSet.<String>of(), zeroFns));

    ...

    "unicode-bidi",
    "var()",
    "vertical-align",
```

But the variable ended up quoted:
```
<div style="color:var( &#39;--text-inactive&#39; );padding-left:10px"></div>
```

**Comment by mikesamuel** [comment]
Why do you have BIT_STRING in there?

**Comment by apottere** [comment]
@mikesamuel because I have no idea what I'm doing.  I tried a bunch of different things that didn't seem to work, though.  Is there a way to do it with the existing code?

**Comment by mikesamuel** [comment]
No.  CssSchema is not really extensible as written.

**Comment by apottere** [comment]
Ok, that's what it seemed like.

**Comment by mikesamuel** [comment]
Yeah.  @apottere The problem is organizational more than technical.

I try to avoid implementing footguns, and have not found a way to allow a highly tuneable definition of "safe CSS" that people outside the CSSWG could reliably use.

HTML is complex but has a reasonably manageable set of tags and attributes, so cautious web designers or developers can probably avoid crafting weak policies by sticking to what they know.  CSS is a much larger, more complicated language, and even seasoned security people get surprised at what you can do with it.

**Comment by jmanico** [comment]
Closing this in an old-issue triage pass. Mike answered it in the thread and that answer still stands, so it is better recorded as closed than left looking unexamined.

The reasoning, from his reply above:

> The problem is organizational more than technical. I try to avoid implementing footguns, and have not found a way to allow a highly tuneable definition of "safe CSS" that people outside the CSSWG could reliably use.

That applies with particular force to custom properties. `--foo: <almost anything>` is valid, a custom property's value is essentially an untyped token stream until some `var()` substitutes it into a real property, and the substitution site is what determines whether the value is dangerous. Vetting `--foo` at the point of declaration means predicting every property it might later land in. A schema flag for "allow CSS variables" would be exactly the kind of switch that looks safe and is not.

Since this was filed, `CssSchema` has become considerably more extensible than it was when Mike wrote "CssSchema is not really extensible as written" — `CssSchema.withProperties`, `union`, and now `toAttributePolicy` let you build and compose a schema without touching the library. So the mechanical objection is largely gone; the design objection about `var()` is what remains.

If you want to revisit it, the productive shape is a concrete proposal for how a `var()` value gets vetted at its substitution site, not a flag to let custom properties through. That is worth a fresh issue.
